drivers/modem/hl7800.c

```diff
@@ -0,0 +1,4887 @@
+	if (ret < 0) {
+		LOG_ERR("AT+COPS ret:%d", ret);
+	}
+
+	/* bring iface up/down */
+	switch (ictx.network_state) {
+	case HL7800_HOME_NETWORK:
+	case HL7800_ROAMING:
+		if (ictx.iface && !net_if_is_up(ictx.iface)) {
+			LOG_DBG("HL7800 iface UP");
+			net_if_up(ictx.iface);
+		}
+		break;
+	case HL7800_OUT_OF_COVERAGE:
+	default:
+		if (ictx.iface && net_if_is_up(ictx.iface)) {
+			LOG_DBG("HL7800 iface DOWN");
+			net_if_down(ictx.iface);
+		}
+		break;
+	}
+
+	if (ictx.iface && !net_if_is_up(ictx.iface)) {
+		hl7800_stop_rssi_work();
+		notify_all_tcp_sockets_closed();
+	} else if (ictx.iface && net_if_is_up(ictx.iface)) {
+		hl7800_start_rssi_work();
+		/* get IP address info */
+		SEND_AT_CMD_IGNORE_ERROR("AT+CGCONTRDP=1");
+		/* get active bands */
+		SEND_AT_CMD_IGNORE_ERROR("AT+KBND?");
+	}
+	LOG_DBG("Network state updated");
+	allow_sleep(true);
+done:
+	hl7800_unlock();
+}
+
+static char *get_network_state_string(enum mdm_hl7800_network_state state)
+{
+	switch (state) {
+		PREFIXED_SWITCH_CASE_RETURN_STRING(HL7800, NOT_REGISTERED);
+		PREFIXED_SWITCH_CASE_RETURN_STRING(HL7800, HOME_NETWORK);
+		PREFIXED_SWITCH_CASE_RETURN_STRING(HL7800, SEARCHING);
+		PREFIXED_SWITCH_CASE_RETURN_STRING(HL7800, REGISTRATION_DENIED);
+		PREFIXED_SWITCH_CASE_RETURN_STRING(HL7800, OUT_OF_COVERAGE);
+		PREFIXED_SWITCH_CASE_RETURN_STRING(HL7800, ROAMING);
+		PREFIXED_SWITCH_CASE_RETURN_STRING(HL7800, EMERGENCY);
+		PREFIXED_SWITCH_CASE_RETURN_STRING(HL7800, UNABLE_TO_CONFIGURE);
+	default:
+		return "UNKNOWN";
+	}
+}
+
+static void set_network_state(enum mdm_hl7800_network_state state)
+{
+	ictx.network_state = state;
+	generate_network_state_event();
+}
+
+static void generate_network_state_event(void)
+{
+	struct mdm_hl7800_compound_event event;
+	event.code = ictx.network_state;
+	event.string = get_network_state_string(ictx.network_state);
+	LOG_INF("Network State: %d %s", ictx.network_state, event.string);
+	event_handler(HL7800_EVENT_NETWORK_STATE_CHANGE, &event);
+}
+
+/* Handler: +CEREG: <n>,<stat>[,[<lac>],[<ci>],[<AcT>]
+ *  [,[<cause_type>],[<reject_cause>] [,[<Active-Time>],[<Periodic-TAU>]]]]
+ */
+static bool on_cmd_network_report_query(struct net_buf **buf, uint16_t len)
+{
+	size_t out_len;
+	char value[MDM_MAX_RESP_SIZE];
+
+	out_len = net_buf_linearize(value, sizeof(value), *buf, 0, len);
+	char *pos = strchr(value, ',');
+	if (pos) {
+		int l = (value + out_len) - pos;
+		char val[l];
+		strncpy(val, pos + 1, l);
+		val[l] = 0;
+		set_network_state(strtol(val, NULL, 0));
+
+		/* start work to adjust iface */
+		k_delayed_work_cancel(&ictx.iface_status_work);
+		k_delayed_work_submit_to_queue(&hl7800_workq,
+					       &ictx.iface_status_work,
+					       IFACE_WORK_DELAY);
+	}
+
+	return true;
+}
+
+#ifdef CONFIG_NEWLIB_LIBC
+/* Handler: +CCLK: "yy/MM/dd,hh:mm:ss±zz" */
+static bool on_cmd_rtc_query(struct net_buf **buf, uint16_t len)
+{
+	struct net_buf *frag = NULL;
+	size_t str_len = strlen(TIME_STRING_FORMAT);
```

Comment:
If you use sizeof(), the str_len will be set at compile time.
So `size_t str_len = sizeof(TIME_STRING_FORMAT) - 1;` would be preferred here.